polar-core/src/vm.rs

```diff
@@ -584,6 +584,28 @@ impl PolarVirtualMachine {
     /// Recursively dereference a variable.
     pub fn deref(&self, term: &Term) -> Term {
         match &term.value() {
+            Value::List(list) => {
+                // Check if last element in list is a rest variable.
+                let mut rest = false;
+                if let Some(last) = list.last() {
+                    if matches!(last.value(), Value::RestVariable(_)) {
+                        rest = true;
+                    }
+                }
+                // Deref all elements.
+                let mut derefed: Vec<Term> = list.iter().map(|t| self.deref(t)).collect();
+                // If last element was a rest variable, append the list it derefed to.
+                if rest {
+                    if let Some(last_term) = derefed.pop() {
+                        if let Value::List(terms) = last_term.value() {
+                            derefed.append(&mut terms.clone());
+                        } else {
+                            derefed.push(last_term);
+                        }
+                    }
+                }
+                term.clone_with_value(Value::List(derefed))
+            }
             Value::Variable(symbol) | Value::RestVariable(symbol) => {
```

Comment:
Does it ever make sense to deref a rest variable outside of the above case with lists? If not, then we could move this arm up into the list case and do something like:

```
if let Some(Value::RestVariable(symbol)) = derefed.last() {
    if let Value::List(terms) = self.value(symbol).value() {
        derefed.pop();
       derefed.append(&mut terms.clone())
   }
```

And you wouldn't need the `rest` logic, nor the case where you pop/push the same term back on.

Reply:
You can definitely deref a rest variable outside derefing a list.
```
query> [1,2,3] = [1,*bs] and x = bs
x => [2, 3]
bs => [2, 3]
```

The push pop logic is needed because the thing the rest variable derefs to could be a list or it could just be the deref variable still (if there's no binding for it yet).
Doing it with a pop/push is simpler than trying to check if it's a list and then doing the pop append because of rusty borrowing reasons.

Reply:
Ah okay. So in my mind that `bs` would be a regular symbol bound to a list? How do we know the `bs` on the RHS is a rest variable?

Reply:
Hmm you might be right, I don't understand your first suggestion though. Once things are derefed it's no longer a list so I don't see how you can avoid doing the check for it being a rest variable before and doing the append after.

Reply:
Oh n/m I get how it works if you don't have the case for RestVariable. I don't know how to test if we need that at the top level thoughl

Reply:
Ok, I tried it like that and it breaks some of these tests, you end up getting non derefed restvariables back as things so I think it has to stay how it is. We could play with simplifying it more later but I'd really like to get this merged since it's broken now.

Reply:
> Ah okay. So in my mind that `bs` would be a regular symbol bound to a list? How do we know the `bs` on the RHS is a rest variable?

I believe that `bs` the normal variable derefs to `*bs` the rest variable which derefs to `[2, 3]` so you can deref a rest variable directly and that's why that version doesn't work.

Reply:
Oh interesting. I find that a little weird. Agree we can revisit later.